File: Testing_Catalog/CAlgorithms/Search/IntTarget/binarySearch.c
//================================================
//================================================
// Binary Search
// binarySearch.c
//
// Isaac A.B. Yep
//
// Time Complexity:
//
// Space Complexity:
//
//================================================
//================================================
#include "main.h"

// algorithm
int binary_search(int* array, int size, int target)
{
    int low = 0;
    int high = size - 1;
    int mid = (low + high) / 2;

    while (low <= high)
    {
        if (array[mid] == target)
            return mid;
        else if (array[mid] < target)
        {
            low = mid + 1;
            mid = (low + high) / 2;
        }
        else if (array[mid] > target)
        {
            high = mid - 1;
            mid = (low + high) / 2;
        }
    }
    return -1;
}

File: Testing_Catalog/Python/Search/IntTarget/binarySearch.py
'''
Binary Search

Runtime Complexity within O(Log2(n))

takes a sorted (smallest to largest) list of integers and integer to find, returns index of target or -1 if not found
'''

def binary_search(array: list, target: int) -> int:
    low = 0
    high = len(array)
    mid = (high + low) // 2
    while low < high:
        mid = (high + low) // 2
        if array[mid] == target:
            return mid
        elif array[mid] < target:
            low = mid + 1
        elif array[mid] > target:
            high = mid
    return -1

